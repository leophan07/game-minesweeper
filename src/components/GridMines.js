import styled, { css } from 'styled-components';

const handleColor = density => {
  switch (parseInt(density)) {
    case 1:
      return "#0000ff";
    case 2:
      return "#cc33ff";
    case 3:
      return "#009900";
    case 4:
      return "#ffaa00";
    case 5:
      return "#8000ff";
    case 6:
      return "#b3b300";
    case 7:
      return "#206020";
    case 8:
      return "#4ce600";
    default:
      return "#cccccc";
  }
};

export const Grid = styled.div`
  width: max-content;
  margin: 0 auto;
  border: 1px solid #aaaaaa;
  display: flex;
  flex-direction: column;
`;

export const Row = styled.div`
  display: flex;
  flex-direction: row;
`;

export const Cell = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #aaaaaa;
  border-radius: 3px;
  margin: 1px;
  width: 30px;
  height: 30px;
  cursor: pointer;
  font-weight: 700;

  ${props => props.isOpen && css`
    background-color: #cccccc ;
    color: ${({ density }) => handleColor(density)};
    cursor: inherit;
  `}
  ${props => props.isOpen && props.isMine && 'background-color: red;'}
`;