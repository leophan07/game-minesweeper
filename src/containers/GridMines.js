import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { setDataSuccess, resetGame } from '../redux/actions';
import { Grid, Row, Cell } from '../components/GridMines';
import { Modal, ModalContent } from '../components/Modal';
import { openCellEmpty, checkWin, msToTime } from '../helpers';
import Button from '../components/Button';
import ButtonGroup from '../components/ButtonGroup';

function GridMinesContainer({ 
  dataGrid, onSetDataSuccess, onResetGame, 
  onPlayGame, isSetting, mines 
}) {
  const [showModal, setShowModal] = useState(null);
  const [startTime, setStartTime] = useState(0);
  const [endTime, setEndTime] = useState(0);

  function openCell(cell) {
    if (startTime === 0) {
      setStartTime(performance.now());
    }

    const { x, y, isMine } = cell;
    if (isMine) {
      setEndTime(performance.now());
      setShowModal('lost');
      openGrid();
    } else {
      let newData = dataGrid;
      newData[x][y].isOpen = true;
      if (checkWin(dataGrid, mines)) {
        setEndTime(performance.now());
        setShowModal('won');
      }
      if (!newData[x][y].density) {
        newData = openCellEmpty(x, y, newData);
      }
      onSetDataSuccess(newData);
    }
  }

  function openGrid() {
    for (let i = 0; i < dataGrid.length; i++) {
      for (let j = 0; j < dataGrid.length; j++) {
        if (!dataGrid[i][j].isOpen && dataGrid[i][j].isMine) {
          dataGrid[i][j].isOpen = true;
        }
      }
    }
    onSetDataSuccess(dataGrid);
  }

  async function onNewGame() {
    await onPlayGame(dataGrid.length);
    setShowModal(null);
    setStartTime(0);
  }

  return (
    <>
      <Grid>
        {dataGrid.map((row, x) => (
          <Row key={x}>
            {row.map((cell, y) => (
              <Cell 
                key={y} 
                isOpen={cell.isOpen} 
                isMine={cell.isMine}
                density={cell.density} 
                onClick={() => openCell(cell)}
              >
                {cell.isOpen ? cell.isMine ? 'ðŸ’£' : cell.density : ''}
              </Cell>
            ))}
          </Row>
        ))}
      </Grid>
      {showModal && 
        <Modal>
          <ModalContent>
            <p>{`You ${showModal} the game in ${msToTime(endTime - startTime)}`}</p>
            <ButtonGroup>
              <Button onClick={onNewGame}>
                {isSetting ? 'Please wait' : 'New Game'}
              </Button>
              <Button onClick={() => onResetGame() && setShowModal(null)}>Home Page</Button>
            </ButtonGroup>
          </ModalContent>
        </Modal>
      }
    </>
  )
}

GridMinesContainer.propTypes = {
  dataGrid: PropTypes.array.isRequired,
  isSetting: PropTypes.bool.isRequired,
  onSetDataSuccess: PropTypes.func.isRequired,
  onResetGame: PropTypes.func.isRequired,
  onPlayGame: PropTypes.func.isRequired,
  mines: PropTypes.number.isRequired,
}

const mapStateToProps = ({ data, isSetting }) => {
  return {
    dataGrid: data,
    isSetting
  };
}

const mapDispatchToProps = dispatch => {
  return {
    onSetDataSuccess: (data) => {
      return dispatch(setDataSuccess(data));
    },
    onResetGame: () => {
      return dispatch(resetGame());
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(GridMinesContainer);