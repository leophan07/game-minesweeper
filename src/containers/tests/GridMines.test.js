import React from 'react';
import GridMines from '../GridMines';
import renderWithRedux from '../../helpers/renderTest';

describe('<GridMinesContainer />', () => {

  it('should render and match the snapshot', () => {
    const { container } = renderWithRedux(
      <GridMines 
        dataGrid={[]}
        onSetDataSuccess={jest.fn()}
        onResetGame={jest.fn()}
        onPlayGame={jest.fn()}
        isSetting={false}
        mines={10}
      />
    );
    expect(container).toMatchSnapshot();
  });
});