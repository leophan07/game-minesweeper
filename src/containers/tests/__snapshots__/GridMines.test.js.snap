// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridMinesContainer /> should render and match the snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa hugSdd"
    data-testid="grid-mines"
  />
</div>
`;
